### Part B. Normalize the data - Regression model
 One hidden layer of 10 nodes, and a ReLU activation function
 Use the adam optimizer and the mean squared error as the loss function.
 Normalizing the data by subtracting the mean from the individual predictors and dividing by the standard deviation. 

In [1]:
import pandas as pd
import numpy as np

### downloading the concrete data

In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

#### Let's check our data points.

In [3]:
concrete_data.shape

(1030, 9)

#### we will check for nulls

In [4]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

In [5]:
from sklearn.model_selection import train_test_split
train_set, test_set = train_test_split(concrete_data,test_size=0.30)
print('train {0}'.format(train_set.shape))
print('test {0}'.format(test_set.shape))

train (721, 9)
test (309, 9)


##### Our target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns. 
##### we need to split the sets up for mse

In [6]:
# need to use train_set and test_set
concrete_data_columns = train_set.columns

predictors_train_set = train_set[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target_train_set = train_set['Strength'] # Strength column
#test set
predictors_test_set = test_set[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target_test_set = test_set['Strength'] # Strength column

#### Normalize the DATA.

In [7]:
#We Normalize the data by subtracting the mean from the individual predictors and dividing by the standard deviation.

predictors_norm_train_set = (predictors_train_set - predictors_train_set.mean()) / predictors_train_set.std()
predictors_norm_test_set = (predictors_test_set - predictors_test_set.mean()) / predictors_test_set.std()

predictors_norm_train_set.head()

num_cols = predictors_norm_train_set.shape[1] # number of predictors
print( num_cols )

8


2. Create the model with one hidden layer. 

In [8]:
import keras
from keras.models import Sequential
from keras.layers import Dense
# define regression model
def regression_model_one_layer():
    # create model with one hidden layer
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(num_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

Using TensorFlow backend.


    2.A Train the model on the training data using 50 epochs.
    3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.
    4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [9]:
# get the list for mse for printing
import statistics 
from sklearn.metrics import mean_squared_error

# build the model
model = regression_model_one_layer()
mse_results=[]
for x in range(50):
    # fit the model
    model.fit(predictors_norm_train_set, target_train_set, epochs=50, verbose=0)
    train_results = model.predict( predictors_norm_test_set )
    mse_results.append( mean_squared_error(target_test_set, train_results) )
    

5. Report the mean and the standard deviation of the mean squared errors.

In [10]:
print("Mean of the list of mean square errors is % s" % (statistics.mean( mse_results ) ))
print("Standard deviation of the list of mean square errors is % s" % ( statistics.stdev( mse_results )))

Mean of the list of mean square errors is 55.430561465455696
Standard deviation of the list of mean square errors is 52.300474600194555


#####################################################################################################################################################################################
# How does the mean of the mean squared errors compare to that from Step A?


### 1) Mean of MSE is greatly reduced compared to part 'A' after using data normalisation. Mean of MSE was 99.73 before normalisation and reduces to 55.430 after data normalisation. 
### 2) Standard deviation is greatly reduced after we use normalized data. The standard deviation of the data was 238.44 before normalisation and reduces to 52.30 after normalisation.